src/main/java/bio/terra/service/resourcemanagement/azure/AzureStorageAccountService.java

```diff
@@ -0,0 +1,325 @@
+import bio.terra.service.resourcemanagement.exception.StorageAccountLockException;
+import bio.terra.service.snapshot.exception.CorruptMetadataException;
+import com.azure.core.management.exception.ManagementException;
+import com.azure.resourcemanager.storage.models.StorageAccount;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Isolation;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Optional;
+import java.util.UUID;
+
+@Component
+public class AzureStorageAccountService {
+    private static final Logger logger = LoggerFactory.getLogger(AzureStorageAccountService.class);
+
+    private final AzureResourceDao resourceDao;
+    private final AzureResourceConfiguration resourceConfiguration;
+    private final ProfileDao profileDao;
+
+    @Autowired
+    public AzureStorageAccountService(
+        AzureResourceDao resourceDao,
+        AzureResourceConfiguration resourceConfiguration,
+        ProfileDao profileDao) {
+        this.resourceDao = resourceDao;
+        this.resourceConfiguration = resourceConfiguration;
+        this.profileDao = profileDao;
+    }
+
+    /**
+     * Fetch metadata for an existing storage_account_resource
+     * Note this method checks for the existence of the underlying cloud resource, if checkCloudResourceExists=true.
+     *
+     * @param storageAccountResourceId id of the storage account resource
+     * @param checkCloudResourceExists true to do the existence check, false to skip it
+     * @return a reference to the storage account as a POJO AzureStorageAccountResource
+     * @throws AzureResourceNotFoundException if no storage_account_resource metadata row is found
+     * @throws CorruptMetadataException if the storage_account_resource metadata row exists but the cloud
+     * resource does not
+     */
+    @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE, readOnly = true)
+    public AzureStorageAccountResource getStorageAccountResourceById(UUID storageAccountResourceId,
+                                                                     boolean checkCloudResourceExists) {
+        // fetch the storage_account_resource metadata row
+        AzureStorageAccountResource storageAccountResource =
+            resourceDao.retrieveStorageAccountById(storageAccountResourceId);
+
+        if (checkCloudResourceExists) {
+            BillingProfileModel billingProfile =
+                profileDao.getBillingProfileById(storageAccountResource.getProfileId());
+            // throw an exception if the storage account does not already exist
+            StorageAccount storageAccount = getCloudStorageAccount(billingProfile, storageAccountResource);
```

Comment:
Really appreciating Phil's comments about returning `Optional`s instead of `null` from a lot of our functions.